pkg/model/iam/iam_builder.go

```diff
@@ -571,13 +602,22 @@ func addECRPermissions(p *Policy) {
 }
 
 func (b *PolicyBuilder) addRoute53Permissions(p *Policy) {
-	// Only the master (unless in legacy mode)
-	if b.Role != kops.InstanceGroupRoleMaster && !b.Cluster.Spec.IAM.Legacy {
+	if b.Role.IsNodeRole() {
+		if featureflag.UseServiceAccountIAM.Enabled() {
+			return
+		}
+
+		// Only the master (unless in legacy mode)
+		if b.Role.NodeRole != kops.InstanceGroupRoleMaster && !b.Cluster.Spec.IAM.Legacy {
+			return
+		}
+	} else if b.Role.ServiceAccountRole != ServiceAccountRoleDNSController {
+		// Only the dns-controller gets route53 permissions
 		return
 	}
 
 	// Legacy IAM permissions for node roles
-	if b.Cluster.Spec.IAM.Legacy {
+	if b.Cluster.Spec.IAM.Legacy && b.Role.IsNodeRole() {
```

Comment:
I would suggest prohibiting having both LegacyIAM and UseServiceAccountIAM enabled at the same time.

Reply:
Done - good call!